File: 8_kyu/Simple_Fun_352_Reagent_Formula.jl
function isvalid(formula)
    material = falses(8)
    for it in formula
        material[it] = true
    end
    ! ( material[1] & material[2] ) &
    ! ( material[3] & material[4] ) &
    ( material[5] == material[6] ) &
    ( material[7] | material[8] )
end

File: 8_kyu/Simple_Fun_352_Reagent_Formula.py
def isValid(formula: list[int]) -> bool:
    f = sum( 1<<bit for bit in formula )
    return (
        f & 0b110 != 0b110 and
        f & 0b11000 != 0b11000 and
        (bit(f,1<<5) == bit(f,1<<6)) and
        (bit(f,1<<7) or bit(f,1<<8)) )
    
def bit(x: int, mask: int) -> bool:
    return x & mask != 0
